src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,41 @@
+package exercise2
+
+import scala.util.Try
+
+trait StackLike[T] {
+  def push(elem: T): StackLike[T]
+
+  def pop(): Try[StackLike[T]]
+
+  def top(): Option[T]
+
+  def isEmpty: Boolean
+
+  def reverse: StackLike[T]
+}
+
+class Stack[T](val elem: Option[T] = None, val tail: Option[Stack[T]] = None)
```

Comment:
A more neat way to implement the stack is to split it into an `EmptyStack` and `Stack`. 
so 
```scala
case class Stack(elem: T, tail:StackLike[T]) extends StackLike[T]
```
and 
```scala
case class EmptyStack[T]() extends StackLike[T]
``` 
or even better
```scala
case object EmptyStack extends StackLike[Nothing]
```
but this has some complications with types, for which we need variance.
I go over this approach briefly in the upcoming lecture.